### DISCUSSION TITLE
UbiquiBot Capabilities Overview

### DISCUSSION THREAD

**Discussion by 0x4007** [ubiquity/ubiquibot#1] [Research]
## Capabilities Overview

Organized by release milestones:

- [x] https://github.com/ubiquity/bounty-bot/discussions/67 - This is required to start pitching to prospective partners.
- [x] https://github.com/ubiquity/bounty-bot/discussions/68 - These are essential capabilities that we need in order to start onboarding prospective partners.
- [ ] https://github.com/ubiquity/bounty-bot/discussions/70 - This automates the financial related capabilities. At this point we should ideally be looking at integrations with our dollar.
- [ ] https://github.com/ubiquity/bounty-bot/discussions/107 - The fancy features that will allow us to promote the project as "AI powered"
- [ ] https://github.com/ubiquity/bounty-bot/discussions/71 - The gamification of the platform. This will likely also require custom UIs (either Chrome Extension injected or a dashboard.)
---
- [ ] https://github.com/ubiquity/bounty-bot/discussions/69 - Communication is miscellaneous and should be regarded as low priority but eventually completed.

**Comment by 0x4007** [comment]
- [x] Import ideas from here https://app.diagrams.net/#G1CVKrfppXfts8vbErmywHHA-TFeC4aJx7

**Comment by 0x4007** [comment]
- It's easier for me to post a new comment when I randomly have an idea and I post from my phone. 

For any new issues created (from these discussions) be sure to link back to the original discussion for the specification of the issue. Do not copy the spec. The reason why is because this is ever-evolving and those specs become stale very quickly relative to the current development velocity (and the updating of spec designs.)

**Comment by 0x4007** [comment]
Sort of a half baked idea but I was skimming over this thread 

https://twitter.com/prtydao/status/1593288065757462530?s=46&t=-bPfKufC5IsS9N-nlRW0-A 

And it inspired me for two things

1. We can brand the bounty system in a more grand/high tech way talking about enabling us all a natively decentralized work future etc 

2. If we could have the community vote by either sending money or just signing a vote to bump up issues/bounties. After quorum the issue can be made into an official bounty.

**Comment by 0x4007** [comment]
I presume that we could save more money on gas fees if we instead only `approve` bounty hunters to `transfer` funds from the multisig. Then, as they see fit, they can claim bounty payments at their preferred pace.

**Reply by cuddle-dev** [reply]
hmm, Even though hunters just need to claim bounty payments, there should be approval action by admin. so i think in terms of gas consumption, we're gonna spend almost same gas.

**Reply by 0x4007** [reply]
Almost != same gas. Any optimizations add up, particularly if we're making hundreds of payments a week.

I suppose that this optimization is so minuscule compared to simply using a rollup so I'm not too aggressively pushing this proposal.

**Reply by 0x4007** [reply]
This could be very interesting to abstract this away with a "cash out" UI for bounty hunters to be able to collect their earnings and pay their own gas fees for it. Its especially useful if they collect tiny bounties like $25-$50 range for example. Then they can roll up those transactions.

**Comment by 0x4007** [comment]
# Meta Bounty Games

We can dynamically promote and remove @ubiquity/bounty-masters based on activity as well. For example, Bounty Masters are expected to open at least one bounty per week or something. That means if a Bounty Hunter starts proposing a ton of bounties (and they are approved) they could be added to the @ubiquity/bounty-masters team dynamically. 

I think for security reasons it would be purely aesthetic and wouldn't really offer more power per se. 

Just an idea to introduce a "leveling system"

Also we can do the same for Bounty Hunters solving bounties at a certain frequency (e.g. one per week) and recognizing that they are "top hunters" etc

**Comment by rndquu** [comment]
@pavlovcik 

Let's breakdown bounty bot features into small tasks.

Not everything is covered but it is enough to start.

We should keep it simple at first because there is a whole bunch of features discussed in the topics above.

## Features

### Share the rules for first time contributors

User stories:
- As I product owner I want to remind first time contributors about the bounty program rules

### Follow up message

Task for this feature: https://github.com/ubiquity/bounty-bot/issues/43

User stories:
- As a product owner I want to understand current development status of an issue

Bot should ask the assigned bounty hunter once in a while about the current progress.

### Weekly update

Task for this feature: https://github.com/ubiquity/bounty-bot/issues/45

User stories:
- As a bounty hunter I want to get a list of weekly updates (to telegram, twitter, etc...) on the finished tasks

### New bounties update

User stories:
- As a bounty hunter I want to receive a weekly update on unresolved bounties

Every Monday bot should get issues which haven't been resolved in at least 7 days and post them in Telegram chat. 

Issue info that should be posted:
- title
- URL
- price

### Assigning bounties

User stories:
- As a bounty hunter I want to be assigned for the task if I have no more questions regarding the task

### Register bounty hunter wallet address

Task for this feature: https://github.com/ubiquity/bounty-bot/issues/42

User stories:
- As a bounty hunter I want to register my wallet address (in order to receive payment) when PR for the issue is merged
- As a product owner I want to track bounty hunter wallet address updates for accounting purposes

When PR is merged the bot should:
- Check the username who created the PR and verify that this PR closed the issue
- Confirm bounty hunter's address

## Tasks to create

### 1. Add first time contributor message

For first time contributors the bot should be able to post a welcome message. 

So when user posts his first message in the repository issues we can consider it a first time contribution.

What should be done:
1. Add parameter `first_time_contributor_message` to the github action config. Example:
```
- name: Ubiquity Bounty Bot
  uses: ubiquity/bounty-bot@v0.0.1
  with: 
    first_time_contibutor_message: Hey, thank you for your first contribution, pls check the rules: https://dao.ubq.fi/devpool
```

If `first_time_contibutor_message` is not set then first time contributor message should not be displayed.

2. When user comments the first time in the repository issues the bot should post the `first_time_contibutor_message` and tag the author. Example:
```
@author Hey, thank you for your first contribution, pls check the rules: https://dao.ubq.fi/devpool
```

3. Update `README.md` with bot config docs.

### 2. Add follow up message

There is already a task for this feature: https://github.com/ubiquity/bounty-bot/issues/43, so perhaps we could update the description.

Bot should ask the assignee about current progress from time to time.

What should be done:
1. Add parameter `follow_up_message` and `follow_up_message_frequency_seconds` to the github action config. Example:
```
- name: Ubiquity Bounty Bot
  uses: ubiquity/bounty-bot@v0.0.1
  with: 
    follow_up_message: Hey, any updates?
    follow_up_message_frequency_seconds: 604800
```

If `follow_up_message` is not set then this message should not be displayed.

Parameter `follow_up_message_frequency_seconds` by default equals to `604800` (7 days)

2. If there are no any comments in the issue for `follow_up_message_frequency_seconds` seconds since the last comment or issue creation then bot should post a follow up message with assignee tag. Example:
```
@author Hey, any updates?
```

3. Update `README.md` with bot config docs.

### 3. Enforce conventional commits

We want the repository commit messages to follow the conventional commits pattern. More info: https://www.conventionalcommits.org/en/v1.0.0/

What should be done:
1. On commit (on a local machine) we should check that commit message follows the conventional commits pattern.
2. Add github action that checks commit messages in PR and fails if they do not follow the conventional commits pattern.

### 4. Implement weekly update report

There is already a task for this feature: https://github.com/ubiquity/bounty-bot/issues/45, so perhaps we could update the description.

The bot should be able to generate a report of resolved issues for the previous week. The report should be posted to the Telegram group https://t.me/UbiquityDevPool.

What should be done:
1. Add parameter `enable_weekly_report` to the github action config. Example: 
```
- name: Ubiquity Bounty Bot
  uses: ubiquity/bounty-bot@v0.0.1
  with: 
    enable_weekly_report: true
```

By default `enable_weekly_report` should be disabled.

2. Every Monday generate weekly report from issues that were resolved during the previous week and post the report to Telegram group https://t.me/UbiquityDevPool. If there were no resolved issues then do not post anything.

Example:
```
Weekly update of resolved issues

- Default Label Colors
- Telegram bot account
- Telegram Bot Adapter
```

3. Update `README.md` with bot config docs.

### 5. Implement unresolved issues report

Every Monday the bot should generate a report for unresolved issues and post the report to Telegram group https://t.me/UbiquityDevPool 

What should be done:
1. Add parameter `enable_unresolved_issues_report` to the github action config. Example:
```
- name: Ubiquity Bounty Bot
  uses: ubiquity/bounty-bot@v0.0.1
  with: 
    enable_unresolved_issues_report: true
```

By default `enable_unresolved_issues_report` should be disabled.

2. Every Moday find all issues that have not been resolved for at least 7 days since their creation, generate a report from them and post to Telegram group https://t.me/UbiquityDevPool. If there were no unresolved issues then do not post anything.

Issue info that should be posted:
- title
- URL
- price

Example:
```
Check our latest unresolved issues

Twitter Bot Account
https://github.com/ubiquity/bounty-bot/issues/56
Pice: 100 USD

CI to RELEASE on merge to main
https://github.com/ubiquity/bounty-bot/issues/52
Price: 100 USD
```

3. Update `README.md` with bot config docs.

### 6. Automatically assign bounties

The bot should be able to automatically assign bounties if bounty hunter has no more questions.

How it should work:
1. Bounty hunter posts his first comment like "I want to take this one" or "I have some questions"
2. The bot answers with:
```
@author Hey, when you have no more questions regarding the issue just type "I want to solve this issue, I don't have any questions" and I will assign the issue to you
```
3. Bounty hunter clears the requiements with bounty masters.
4. Bounty hunter comments "I want to solve this issue, I don't have any questions" and the bot assigns the issue

What should be done:
1. Add the following parameters to the github action config:
    - `auto_assign_confirmation_message`: Text message used by the bounty hunter to notify the bot that bounty hunter has no more questions and the issue can be assigned. By default is empty. If it is empty the auto assign feature should not work.
    - `auto_assign_exclude_teams`: The bot will not post his introductory message "@author Hey, when you have no more questions..." to users from those teams. It is useful to prevent bot spamming when multiple bounty masters comment on the issue. If bounty master want to take the issue himself he could easily type confirmation message as a normal bounty hunter. By default an empty array.
    - `auto_assign_max_unresolved_issues`: The number of max unresolved issues. There is a typical situation when bounty hunter posts a PR for the issue and wants to take the next bounty while PR is in review. We should allow him to take a new one if he has no more than `auto_assign_max_unresolved_issues` number of issues in review. By default equals to 1.

Example:
```
- name: Ubiquity Bounty Bot
  uses: ubiquity/bounty-bot@v0.0.1
  with: 
    auto_assign_confirmation_message: I want to solve this issue, I don't have any questions
    auto_assign_exclude_teams: [@ubiquity/development, @ubiquity/bounty-masters]
    auto_assign_max_unresolved_issues: 1
```

2. Implement the bot with the following features:
- From time to time monitor comments in the issues
- If there is a new comment (check that this is user's 1st comment in the issue) in the issue from the user who is not in the `auto_assign_exclude_teams`, then post a message:
```
@author Hey, when you have no more questions regarding the issue just type "I want to solve this issue, I don't have any questions" and I will assign the issue to you
```
- If user types `auto_assign_confirmation_message` then check that he has no more than `auto_assign_max_unresolved_issues` number of unresolved issues. If max number is not reached then the bot should assign user to the issue. If max number is reached then bot should post an error message:
```
@author Sorry, max number of unresolved issues reached
```

3. Update `README.md` with bot config docs.

### 7. Register wallet address

There already exists the task for this feature: Task for this feature: https://github.com/ubiquity/bounty-bot/issues/42. Perhaps we could update the description.

Depends on:
- https://github.com/ubiquity/bounty-bot/issues/46

The bot should be able to register (save in the DB) bounty hunter's wallet address when PR is merged. Bounty hunter can post different wallet addresses for different issues so we should track wallet address updates for accounting purposes.

When PR is merged the bot should:
- Check the username who created the PR and verify that this PR closed the issue
- Confirm bounty hunter's address

What should be done:
1. Add the following parameters to the github action config:
    - `register_wallet_ask_message`: This message the bot posts when PR is merged asking for the wallet address. By default is empty. If it is empty then "register wallet" feature should not work.
    - `register_wallet_confirmation_message`: This message contains wallet adddress. It is posted by bounty hunter in response to the `register_wallet_ask_message`.
    - `register_wallet_success_message`: This message is posted when the bot saves bounty hunter's wallet address into the DB.

Example:
```
- name: Ubiquity Bounty Bot
  uses: ubiquity/bounty-bot@v0.0.1
  with: 
    register_wallet_ask_message: Pls share your wallet address
    register_wallet_confirmation_message: My wallet address is
    register_wallet_success_message: Saved it, thank you
```

2. Implement the bot with the following features:
- When PR for the issue is merged then check that PR authors equals to issue assignee. If they are not equals then post an error with issue assignee tag:
```
@author Sorry, PR author and issue assignee do not match
```
- If PR author and issue assignee match then bot posts the `register_wallet_ask_message` with issue assignee tag, example:
```
@author Pls share your wallet address
```
- The bot monitors issue comments for `register_wallet_confirmation_message` pattern. If comment author equals to issue assignee then the bot saves the wallet address to the DB. Example response from the bounty hunter:
```
My wallet address is 0xe57c2c8839ea4510ac0fee8ce9f1435e0a1f992e
```
- The bot validates wallet address. If it is not valid then the bot posts an error with issue assignee tag:
```
@author Sorry, invalid wallet address
```
- If wallet address is valid then the bot saves it to the DB. 

You should also update the DB schema and adapters (supabase calls) for this task in order to track bounty hunter wallet updates. 

The DB schema may look like this:

**user**
| id | github_username |
| -------- | -------- |
| Number     | Text     |

**wallet**
| id | address |
| -------- | -------- |
| Number     | Text     |

**user_wallet**
| id | user_id | wallet_id |
| -------- | -------- | ------- |
| Number     | Number     | Number |

For DB adapter pls follow this task: https://github.com/ubiquity/bounty-bot/issues/46

- When the bot saves wallet address it posts the `register_wallet_success_message` with issue assignee tag. Example:
```
@author Saved it, thank you
```

3. Update `README.md` with bot config docs.

**Reply by rndquu** [reply]
@pavlovcik pls review again and say if everything is fine

@0xcodercrane said it looks nice and proposed to use default messages for all strings (like `first_time_contributor_message`, `follow_up_message`, etc...) if they are not set

if everything looks fine then I'll create (or update existing tasks) and share the results with you and @0xcodercrane to check

**Reply by 0x4007** [reply]
Apologies for missing your write-up! 

My concern is that these specs are constantly evolving which is why I think it makes sense to make the bounty/issue official right before implementation. @0xcodercrane and I had some discussion around this but some slipped through. Moving forward, lets be sure to just link back to the discussion instead of copying the details on the issue. That way we can keep the latest idea synchronized.

Below is an example of what I mean by "linking back" instead of copying the spec.

<img width="1188" alt="image" src="https://user-images.githubusercontent.com/4975670/215267511-319f58c4-c1af-4a7c-9f34-3d4895269043.png">

Something like this but I'm sure you can figure something nicer.

---

It's a bit difficult to reply due to how large your comment is. I think it makes more sense to comment on each issue with your thoughts. If you must create a new issue, I think either just create the issue (and we can discuss there) or compile into this discussion. It'll help break up the conversation and make it more manageable.

> Weekly update

I also want to add that its used to ping the group to remind them that our project exists. I'm hesitant on the bounty at this stage because we couldn't possibly have that for every partner project in a single group chat! It's just super nice-to-have to potentially motivate the bounty hunters too. Perhaps this could be broken off into a separate script just for our project (and not partner projects?)

> New bounties update

This is another idea that I fear will not scale with when we onboard partner projects. This falls in the same category as above.

> Assigning bounties

I added detail about removing the need to support stateful conversations for this here. Its essentially sending a keyword/command to the bot to assign them so that they get "dibs" on the bounty. https://github.com/ubiquity/bounty-bot/discussions/67

> Register bounty hunter wallet address

I updated #42

> 1. Add first time contributor message

Its a great idea. Make an issue and we can refine it if needed.

> 2. Add follow up message

Also cool idea. I'm not confident on how long we should wait, currently I feel that "4-7 days" is ok but I'm not convinced that its precise enough for production. A bit off-topic but it could be interesting to have another config that refers to "development velocity" where if it is "war time" and we need to ship super fast, we can set the dev velocity to its max setting and vice versa. This could affect parameters like follow up time (daily->weekly->monthly etc) and some other details I haven't thought about yet. We could even automate where, the higher the price of ETH, the higher the pressure (and base pricing) should be to ship faster lol

> 3. Enforce conventional commits

Cool

> 4. Implement weekly update report

Seems like you wrote about it above and I already provided my response there. Since we have to train a GPT model, I think we should backlog it.

> 5. Implement unresolved issues report

I think the terminology I used was "stale" issues. Could be interesting to bring them to light. I also think at the same time we should boost the price by 5% every week. Eventually all stale issues should get done if the price is high enough.

> 6. Automatically assign bounties

I updated this discussion with an idea for a command to get assigned. I disagree with the bot being so chatty. It should only speak when the command is issued by the bounty hunter to be assigned the issue.

> `auto_assign_max_unresolved_issues` is an interesting one. I noticed that I'm strict with new talent only having one, but proven talent I'm okay with them having several assigned at the same time. Also when the delay is on our side with reviews, I am ok to assign bounty hunters another bounty while they wait for us. 

> 7. Register wallet address

I updated this spec.

> DB Schema

We have a ton of data not included in your proposal here so I'll disregard for now. I can share the Supabase instance sometime so you can see what we're working with.

**Comment by 0x4007** [comment]
An interesting incentive for pull request reviewers to do a good job could be a delayed permit to be generated. 

For example, if a pull request is not reverted in one month (configurable) the bot will generate a “high quality review award” otherwise if the pull request I reverted for any reason (e.g. a bug was found) then the permit will not be generated and a warning message will explain why. 

It could be structured as a “half up front, half later” payout for reviewers. 

Sometimes a revert might be the best action to indicate a contested payout/review so we could always consider a command from an admin toggling the payout generation.
